### ISSUE TITLE
Metrics column being generated without collector key prefix

### ISSUE THREAD

**Issue by MrFallen** [HugoJF/csgo-sessions#3]
![image](https://user-images.githubusercontent.com/5339386/84486251-7ee7e580-ace0-11ea-8eca-ad422e286ddf.png)
![image](https://user-images.githubusercontent.com/5339386/84486269-860ef380-ace0-11ea-9d17-cec817a1220c.png)

Can u help me? :(

**Comment by MrFallen** [comment]
![image](https://user-images.githubusercontent.com/5339386/84548248-9064d800-ad48-11ea-91fe-36f74ba52510.png)

**Comment by HugoJF** [comment]
If the `metrics` column is filled, then Pipeline is working fine.

  - Can you please provide the output of `/sessions/11/raw`? It would output what comes out of the `SessionBuilder`
  - Can you provide the JSON metrics from session 11?
  - Are there any errors on ` storage/logs/`?

**Comment by MrFallen** [comment]
@HugoJF hello!

> * Can you please provide the output of `/sessions/11/raw`?

![image](https://user-images.githubusercontent.com/5339386/84566477-cc358700-adb4-11ea-85e4-19d6385293c1.png)

> * Can you provide the JSON metrics from session 11?

`{"ak47.right_leg":"26","deagle.stomach":"47","ak47.chest":"78","ak47.head":"247","sg556.chest":"56","sg556.right_arm":"28","ak47.right_arm":"1","mp9.head":"81","deagle.head":"156","mp9.chest":"1"}`

> * Are there any errors on ` storage/logs/`?

No. Logs are clear like this:
`[2020-06-13 09:52:04] local.INFO: Processing took: 0.01 seconds to process 0 events [10.0 ms/event]  
[2020-06-13 09:53:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 09:53:03] local.INFO: Processing took: 0.03 seconds to process 0 events [30.0 ms/event]  
[2020-06-13 09:54:04] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 09:54:04] local.INFO: Processing took: 0.04 seconds to process 0 events [40.0 ms/event]  
[2020-06-13 09:55:09] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 09:55:09] local.INFO: Processing took: 0.03 seconds to process 0 events [30.0 ms/event]  
[2020-06-13 09:56:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 09:56:03] local.INFO: Processing took: 0.03 seconds to process 0 events [30.0 ms/event]  
[2020-06-13 09:57:05] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 09:57:05] local.INFO: Processing took: 0.03 seconds to process 0 events [30.0 ms/event]  
[2020-06-13 09:58:04] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 09:58:04] local.INFO: Processing took: 0.03 seconds to process 0 events [30.0 ms/event]  
[2020-06-13 09:59:04] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 09:59:04] local.INFO: Processing took: 0.04 seconds to process 0 events [40.0 ms/event]  
[2020-06-13 10:00:08] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:00:08] local.INFO: Processing took: 0.03 seconds to process 0 events [30.0 ms/event]  
[2020-06-13 10:00:09] local.INFO: Found 0 stale sessions {"sessions":{"Illuminate\\Support\\Collection":[]}} 
[2020-06-13 10:01:05] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:01:05] local.INFO: Processing took: 0.11 seconds to process 0 events [110.0 ms/event]  
[2020-06-13 10:02:04] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:02:04] local.INFO: Processing took: 0.09 seconds to process 0 events [90.0 ms/event]  
[2020-06-13 10:03:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:03:03] local.INFO: Processing took: 0.04 seconds to process 0 events [40.0 ms/event]  
[2020-06-13 10:04:04] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:04:04] local.INFO: Processing took: 0.06 seconds to process 0 events [60.0 ms/event]  
[2020-06-13 10:05:07] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:05:07] local.INFO: Processing took: 0.04 seconds to process 0 events [40.0 ms/event]  
[2020-06-13 10:06:02] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:06:02] local.INFO: Processing took: 0.01 seconds to process 0 events [10.0 ms/event]  
[2020-06-13 10:07:04] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:07:04] local.INFO: Processing took: 0.08 seconds to process 0 events [80.0 ms/event]  
[2020-06-13 10:08:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:08:03] local.INFO: Processing took: 0.10 seconds to process 0 events [100.0 ms/event]  
[2020-06-13 10:09:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:09:03] local.INFO: Processing took: 0.02 seconds to process 0 events [20.0 ms/event]  
[2020-06-13 10:10:06] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:10:06] local.INFO: Processing took: 0.04 seconds to process 0 events [40.0 ms/event]  
[2020-06-13 10:11:05] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:11:05] local.INFO: Processing took: 0.06 seconds to process 0 events [60.0 ms/event]  
[2020-06-13 10:12:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:12:03] local.INFO: Processing took: 0.02 seconds to process 0 events [20.0 ms/event]  
[2020-06-13 10:13:04] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:13:04] local.INFO: Processing took: 0.03 seconds to process 0 events [30.0 ms/event]  
[2020-06-13 10:14:02] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:14:02] local.INFO: Processing took: 0.03 seconds to process 0 events [30.0 ms/event]  
[2020-06-13 10:15:06] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:15:06] local.INFO: Processing took: 0.02 seconds to process 0 events [20.0 ms/event]  
[2020-06-13 10:15:06] local.INFO: Found 0 stale sessions {"sessions":{"Illuminate\\Support\\Collection":[]}} 
[2020-06-13 10:16:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:16:03] local.INFO: Processing took: 0.02 seconds to process 0 events [20.0 ms/event]  
[2020-06-13 10:17:04] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:17:04] local.INFO: Processing took: 0.07 seconds to process 0 events [70.0 ms/event]  
[2020-06-13 10:18:02] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:18:02] local.INFO: Processing took: 0.04 seconds to process 0 events [40.0 ms/event]  
[2020-06-13 10:19:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:19:03] local.INFO: Processing took: 0.06 seconds to process 0 events [60.0 ms/event]  
[2020-06-13 10:20:08] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:20:08] local.INFO: Processing took: 0.05 seconds to process 0 events [50.0 ms/event]  
[2020-06-13 10:21:04] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:21:04] local.INFO: Processing took: 0.08 seconds to process 0 events [80.0 ms/event]  
[2020-06-13 10:22:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:22:03] local.INFO: Processing took: 0.05 seconds to process 0 events [50.0 ms/event]  
[2020-06-13 10:23:02] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:23:02] local.INFO: Processing took: 0.03 seconds to process 0 events [30.0 ms/event]  
[2020-06-13 10:24:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:24:03] local.INFO: Processing took: 0.02 seconds to process 0 events [20.0 ms/event]  
[2020-06-13 10:25:07] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:25:07] local.INFO: Processing took: 0.02 seconds to process 0 events [20.0 ms/event]  
[2020-06-13 10:26:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:26:03] local.INFO: Processing took: 0.02 seconds to process 0 events [20.0 ms/event]  
[2020-06-13 10:27:02] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:27:02] local.INFO: Processing took: 0.02 seconds to process 0 events [20.0 ms/event]  
[2020-06-13 10:28:02] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:28:02] local.INFO: Processing took: 0.01 seconds to process 0 events [10.0 ms/event]  
[2020-06-13 10:29:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:29:03] local.INFO: Processing took: 0.01 seconds to process 0 events [10.0 ms/event]  
[2020-06-13 10:30:07] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:30:07] local.INFO: Processing took: 0.02 seconds to process 0 events [20.0 ms/event]  
[2020-06-13 10:30:07] local.INFO: Found 0 stale sessions {"sessions":{"Illuminate\\Support\\Collection":[]}} 
[2020-06-13 10:30:08] local.INFO: Skipped initialization of App\Stats\Kills\TotalKillsStat (already loaded)  
[2020-06-13 10:30:08] local.INFO: Skipped initialization of App\Stats\Kills\KillsPerHitGroupStat (already loaded)  
[2020-06-13 10:30:08] local.INFO: Skipped initialization of App\Stats\Kills\TotalKillsStat (already loaded)  
[2020-06-13 10:30:08] local.INFO: Skipped initialization of App\Stats\Deaths\TotalDeathsStat (already loaded)  
[2020-06-13 10:30:08] local.INFO: Skipped initialization of App\Stats\Kills\TotalKillsStat (already loaded)  
[2020-06-13 10:30:08] local.INFO: Skipped initialization of App\Stats\Misc\DurationStat (already loaded)  
[2020-06-13 10:30:08] local.INFO: Skipped initialization of App\Stats\Damage\TotalDamageStat (already loaded)  
[2020-06-13 10:30:08] local.INFO: Skipped initialization of App\Stats\Misc\DurationStat (already loaded)  
[2020-06-13 10:31:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:31:03] local.INFO: Processing took: 0.03 seconds to process 0 events [30.0 ms/event]  
[2020-06-13 10:32:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:32:03] local.INFO: Processing took: 0.02 seconds to process 0 events [20.0 ms/event]  
[2020-06-13 10:33:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:33:03] local.INFO: Processing took: 0.08 seconds to process 0 events [80.0 ms/event]  
[2020-06-13 10:34:03] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:34:03] local.INFO: Processing took: 0.03 seconds to process 0 events [30.0 ms/event]  
[2020-06-13 10:35:08] local.INFO: Null event found {"rawEvent":false,"event":null} 
[2020-06-13 10:35:08] local.INFO: Processing took: 0.06 seconds to process 0 events [60.0 ms/event]  
`

**Comment by HugoJF** [comment]
Well, after thinking what could be happening and writing a bunch of considerations for when I have more time to debug this, I realized that the `metrics` JSON is just invalid. Non of them are prefixed correctly.

Here is what a valid JSON looks like:

```json
{"damage-received.awp.stomach":"550","damage-received.ak47.left_arm":"92","damage.ak47.right_arm":"187","deaths.awp.head":"1","kills.ak47.chest":"4","damage-received.ak47.neck":"27","kills.m4a1.right_arm":"2","damage-received.awp.left_leg":"85","damage-received.deagle.chest":"48","damage-received.fiveseven.chest":"21","damage-received.awp.head":"447","deaths.ak47.chest":"6","damage-received.ak47.right_leg":"90","damage.ak47.left_leg":"129","damage.m4a1.left_arm":"44","damage-received.deagle.right_arm":"11","hits.ak47.right_leg":"5","hits.ak47.chest":"43","hits.ak47.left_leg":"5","damage-received.deagle.head":"636","damage.m4a1.right_leg":"44","hits.m4a1.head":"18","damage-received.ak47.left_leg":"52","damage-received.ak47.right_arm":"248","damage.ak47.left_arm":"78","deaths.ak47.left_arm":"1","damage-received.ak47.head":"3271","deaths.awp.chest":"6","deaths.deagle.head":"4","kills.m4a1.head":"8","damage.ak47.head":"1997","deaths.ak47.stomach":"8","hits.m4a1.chest":"58","deaths.awp.stomach":"4","damage-received.ak47.stomach":"901","hits.ak47.left_arm":"3","damage.ak47.right_leg":"113","hits.m4a1.right_leg":"2","damage.m4a1.stomach":"372","kills.ak47.left_arm":"1","deaths.ak47.head":"31","hits.m4a1.right_arm":"9","damage-received.awp.chest":"658","hits.ak47.head":"19","damage.m4a1.chest":"1232","deaths.ak47.right_arm":"2","kills.m4a1.stomach":"2","kills.m4a1.chest":"12","hits.ak47.stomach":"22","damage-received.ak47.chest":"1303","hits.ak47.right_arm":"7","damage.m4a1.head":"1539","kills.ak47.head":"19","kills.m4a1.right_leg":"1","damage.ak47.chest":"1144","hits.m4a1.left_arm":"2","kills.ak47.stomach":"2","kills.ak47.right_leg":"1","damage.ak47.stomach":"705","hits.m4a1.stomach":"14","deaths.ak47.left_leg":"1","damage.m4a1.right_arm":"196"}
```

Can you provide me with the return value of `KEYS *` from your REDIS server? This should help me pinpoint if the issue is during the session (maybe the keys used on REDIS) or when it's closing the session and building the `metrics` JSON.

EDIT: I think I know what's happening: you probably don't the `REDIS_PREFIX` variable set, and that would make the session `metrics` builder function to fuck the stats type prefix.

Instead of storing variables as `csgosessions.11.damage.ak47.head` it probably storing it as `11.damage.ak47.head`. Since I'm using the Redis facade to query anything starting with `11`, it works as expected (ignoring any prefixes). But what it returns me is the full key (without the prefix), while [this guy](https://github.com/HugoJF/csgo-sessions/blob/master/app/Services/SessionService.php#L35) expects a prefix. Because of that, the regex removes the ID and the stat type (instead of prefix and ID), causing everything to fuck up.

TL;DR: add `REDIS_PREFIX=csgosessions.` to your `.env` file. That should fix everything.

**Comment by MrFallen** [comment]
@HugoJF I love u dude. Thats works!
